src/Rewards.sol

```diff
@@ -0,0 +1,181 @@
+     */
+    uint256 public quotaRenewalTimestamp;
+
+    /**
+     * @dev Amount of rewards claimed in the current period.
+     */
+    uint256 public rewardsClaimed;
+
+    /**
+     * @dev Address of the authorized oracle.
+     */
+    address public authorizedOracle;
+
+    /**
+     * @dev Mapping to store reward counters for each recipient to prevent replay attacks.
+     */
+    mapping(address => uint256) public rewardCounters;
+
+    /**
+     * @dev Struct on which basis an individual reward must be issued.
+     */
+    struct Reward {
+        address recipient;
+        uint256 amount;
+        uint256 counter;
+    }
+
+    /**
+     * @dev Error when the reward quota is exceeded.
+     */
+    error RewardQuotaExceeded();
+
+    /**
+     * @dev Error when the reward is not from the authorized oracle.
+     */
+    error UnauthorizedOracle();
+
+    /**
+     * @dev Error when the recipient's counter does not match.
+     */
+    error InvalidRecipientCounter();
+
+    /**
+     * @dev Event emitted when the reward quota is set.
+     */
+    event RewardQuotaSet(uint256 quota);
+
+    /**
+     * @dev Event emitted when a reward is minted.
+     */
+    event RewardMinted(address indexed recipient, uint256 amount, uint256 totalRewardsClaimed);
+
+    /**
+     * @dev Initializes the contract with the specified parameters.
+     * @param nodlTokenAddress Address of the NODL token contract.
+     * @param initialQuota Initial reward quota.
+     * @param initialPeriod Initial reward period.
+     * @param oracleAddress Address of the authorized oracle.
+     */
+    constructor(address nodlTokenAddress, uint256 initialQuota, uint256 initialPeriod, address oracleAddress)
+        EIP712(SIGNING_DOMAIN, SIGNATURE_VERSION)
+    {
+        _grantRole(DEFAULT_ADMIN_ROLE, msg.sender);
+        nodlToken = NODL(nodlTokenAddress);
+        rewardQuota = initialQuota;
+        rewardPeriod = initialPeriod;
+        rewardsClaimed = 0;
+        quotaRenewalTimestamp = block.timestamp + rewardPeriod;
+        authorizedOracle = oracleAddress;
+    }
+
+    /**
+     * @dev Sets the reward quota. Only accounts with the QUOTA_SETTER_ROLE can call this function.
+     * @param newQuota The new reward quota.
+     */
+    function setRewardQuota(uint256 newQuota) external {
+        _checkRole(QUOTA_SETTER_ROLE);
+        rewardQuota = newQuota;
+        emit RewardQuotaSet(newQuota);
+    }
+
+    /**
+     * @dev Mints rewards to the recipient if the signature is valid and quota is not exceeded.
+     * @param reward The reward details.
+     * @param signature The signature from the authorized oracle.
+     */
+    function mintReward(Reward memory reward, bytes memory signature) external {
+        bytes32 digest = _hashTypedDataV4(
+            keccak256(abi.encode(keccak256(REWARD_TYPE), reward.recipient, reward.amount, reward.counter))
+        );
+        address signer = ECDSA.recover(digest, signature);
```

Comment:
Using [`SignatureChecker`](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/utils/cryptography/SignatureChecker.sol) instead might allow us a bit more flexibility if we want to support more complex signature schemes later. To be discussed

Reply:
Implemented by https://github.com/NodleCode/rollup/pull/31/commits/77ca861356d3a75e122505754a62e55cb3741e50
Thank you for highlighting it for me.